[numpy] Add  torch.take_along_dim

kshitij12345 opened the pull request:
Reference: #38349 

Wrapper around the existing `torch.gather` with broadcasting logic.

TODO:
* [x] Add Doc entry (see if phrasing can be improved)
* [x] Add OpInfo
* [x] Add test against numpy
* [x] Handle broadcasting behaviour and when dim is not given.

kshitij12345 committed:
add function for take_along_dim

kshitij12345 committed:
fix interned_string name

kshitij12345 committed:
add overries entry

kshitij12345 committed:
add opinfo test based on gather

kshitij12345 committed:
update doc entries and signature

kshitij12345 committed:
fix typo

kshitij12345 committed:
update signature and broadcasting logic

kshitij12345 committed:
remove invalid entries

kshitij12345 committed:
Merge branch 'master' into develop/numpy/add-take_along_dim

kshitij12345 committed:
update logic for broadcasting

kshitij12345 committed:
add test

kshitij12345 committed:
update signature and replace take path with gather

kshitij12345 committed:
update test and samples for test

kshitij12345 committed:
refactor and comment

kshitij12345 committed:
update doc phrasing

kshitij12345 committed:
add out in documented signature

kshitij12345 committed:
update overrides.py signature

mruberry commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -1162,6 +1162,60 @@ Tensor take_backward(const Tensor& grad, const Tensor& input, const Tensor& inde
+    int64_t dim) {
+  TORCH_CHECK(
+      self.dim() == indices.dim(),
+      "self and indices should have same number of dimensions")
```
"torch.take_along_dim(): input and indices should have the same number of dimensions, but got X dimensions for input, and Y dimensions for indices"

Remember to start with the name of the function and that we use "input" instead of "self" in our user-facing documentation (which is confusing, I know).

mruberry replied:
Also check that the tensors are on the same device. Is there any dtype validation that should occur, too?

We should get in the habit of thinking: device, dtype, shape checking when implementing operators. (I sometimes forget, too!)

kshitij12345 replied:
For `indices`, we verify the `dtype` to be `Long`.

For `device`, I'll add a test to confirm. 

Thanks!

mruberry replied:
Adding the device check upfront is OK because it will provide a clearer error message. While we could rely on gather's device checking, if the device check is tripped in gather() then the error message should say "torch.gather(): ...", which would be more confusing to users since it's not the function they called.

kshitij12345 replied:
I'll add a check upfront.

However at the moment,
for `torch.gather`,  check for all tensors being on same device is carried out by TensorIterator and it prints the message without the function name. (Maybe in the redesign of TensorIterator it should also take the API name to print it in warnings and errors).

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5744,6 +5744,13 @@
   device_guard: False
 
+- func: take_along_dim.out(Tensor self, Tensor indices, int? dim=None, *, Tensor(a!) out) -> Tensor(a!)
+  dispatch:
```
Wouldn't this be a MATH dispatch because it's a composite?

mruberry commented on test/test_indexing.py:
```diff
@@ -1041,6 +1043,56 @@ def test_invalid_device(self, device):
+            self.assertEqual(actual, expected, atol=0, rtol=0)
+
+        for shape in [(2, 3, 5), (3, 2), (2, 3, 1, 4)]:
+            t = torch.randn(shape, device=device)
```
use make_tensor()

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8810,6 +8810,33 @@ def merge_dicts(*dicts):
+           r"""
+take_along_dim(input, indices, dim, *, out=None) -> Tensor
+
+Returns values from the :attr:`input` at the given 1-d indices along the given dim.
```
"Returns" -> "Selects"

mruberry replied:
"at the 1-dimensional indices from :attr:\`indices\` along the given :attr:\`dim\`."

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8810,6 +8810,33 @@ def merge_dicts(*dicts):
+
+Args:
+    {input}
+    indices (LongTensor): the indices into tensor
```
indices (tensor): the indices into :attr:\`input\`. Must have long dtype.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8810,6 +8810,33 @@ def merge_dicts(*dicts):
+Args:
+    {input}
+    indices (LongTensor): the indices into tensor
+    dim (int): dim to take 1d slices along.
```
"dimension to select along"

Let's not use terms like "slice" because they're not defined.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8810,6 +8810,33 @@ def merge_dicts(*dicts):
+
+    >>> a = torch.tensor([[10, 30, 20], [60, 40, 50]])
+    >>> sorted_idx = torch.tensor([[0, 2, 1], [1, 2, 0]])
+    >>> torch.take_along_dim(a, sorted_idx, dim=1)
```
Would you update the docs for argmin and argmax, too? I think they should be paired with this function.

See https://github.com/pytorch/pytorch/pull/49684#discussion_r546834844 and the NumPy documentation.

mruberry commented on test/test_indexing.py:
```diff
@@ -1041,6 +1043,56 @@ def test_invalid_device(self, device):
+            expected = np.take_along_axis(t_np, indices_np, axis=dim)
+            self.assertEqual(actual, expected, atol=0, rtol=0)
+
+        for shape in [(2, 3, 5), (3, 2), (2, 3, 1, 4)]:
```
Are there other degenerate but valid inputs this should check? Like what about tensors with no elements, and noncontiguous tensors (make_tensor() has a "discontiguous" kwarg now to make them easy to test)?

mruberry commented on test/test_indexing.py:
```diff
@@ -1041,6 +1043,56 @@ def test_invalid_device(self, device):
+            _test_against_numpy(t, indices, dim)
+
+        # dim of `t` and `indices` does not match
+        with self.assertRaises(RuntimeError):
```
Separate error checks into a separate test

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -635,6 +635,35 @@ def sample_inputs_gather(op_info, device, dtype, requires_grad):
                         0, torch.tensor(0, dtype=torch.int64, device=device))),
             )
 
+def sample_inputs_take_along_dim(op_info, device, dtype, requires_grad):
```
How long do these new tests take in the CI?

kshitij12345 replied:
```
pytest test/test_ops.py -k _take_along_dim
========================================================================= test session starts ==========================================================================
platform linux -- Python 3.8.6, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /home/kshiteej/Pytorch/pytorch_take_along_axis
plugins: hypothesis-5.38.1
collected 9469 items / 9360 deselected / 109 selected                                                                                                                  

test/test_ops.py ............................ssss....ssss.....................................................................                                   [100%]
===================================================== 101 passed, 8 skipped, 9360 deselected, 4 warnings in 24.15s =====================================================
```

mruberry replied:
@albanD we really need to reduce our grad{grad, }check time.

kshitij12345 replied:
For referece,

Durations with `pytest --durations`

cmd : `pytest test/test_ops.py -k _take_alo --durations=30`

Output
```
1.27s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_take_along_dim_cuda_bfloat16
1.10s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_take_along_dim_cuda_complex128
0.70s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_take_along_dim_cuda_complex128
0.67s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_take_along_dim_cuda_complex64
0.66s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_take_along_dim_cuda_float32
0.64s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_take_along_dim_cuda_float16
0.62s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_take_along_dim_cuda_float64
0.54s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_take_along_dim_cuda_bfloat16
0.49s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_take_along_dim_cuda_float64
0.41s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_take_along_dim_cuda_complex128
0.34s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_take_along_dim_cpu_float64
0.34s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_take_along_dim_cuda_uint8
0.34s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_take_along_dim_cpu_float32
0.33s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_take_along_dim_cuda_int8
0.33s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_take_along_dim_cuda_int16
0.32s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_take_along_dim_cuda_int32
0.32s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_take_along_dim_cpu_complex64
0.32s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_take_along_dim_cpu_complex128
0.31s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_take_along_dim_cuda_int64
0.30s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_take_along_dim_cpu_float16
0.30s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_take_along_dim_cuda_bool
0.26s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_take_along_dim_cuda_bool
0.26s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_take_along_dim_cpu_complex128
0.26s call     test/test_ops.py::TestCommonCUDA::test_out_take_along_dim_cuda_bfloat16
0.25s call     test/test_ops.py::TestCommonCUDA::test_out_take_along_dim_cuda_bool
0.24s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_take_along_dim_cuda_float64
0.23s call     test/test_ops.py::TestCommonCUDA::test_out_take_along_dim_cuda_complex128
0.23s call     test/test_ops.py::TestCommonCUDA::test_out_take_along_dim_cuda_float32
0.23s call     test/test_ops.py::TestCommonCUDA::test_out_take_along_dim_cuda_complex64
0.23s call     test/test_ops.py::TestCommonCUDA::test_out_take_along_dim_cuda_float16
===================================================== 101 passed, 8 skipped, 9781 deselected, 4 warnings in 24.91s =====================================================
```

`1.27s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_take_along_dim_cuda_bfloat16` takes long probably because that is where CUDA is initialized first time.

mruberry replied:
Thank you, @kshitij12345. I didn't realize the jit tests were often so slow. Maybe (not in this PR) we should consider pruning their dtypes even more.

mruberry reviewed (commented):
Hey @kshitij12345! It's great to finally have a torch.take_along_dim() implementation! Overall this looks good, as usual. I made a few suggestions and asked a few questions for you to review.

kshitij12345 committed:
address comments

* add test for cross-device input
* seperate test for invalid inputs
* update error messages
* update dispatch to Math
* update docs

kshitij12345 committed:
upfront check for different devices in take_along_dim

kshitij12345 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -1162,6 +1162,81 @@ Tensor take_backward(const Tensor& grad, const Tensor& input, const Tensor& inde
+  return std::make_tuple(self_broadcasted, indices_broadcasted, dim);
+}
+
+static inline void checkDevice(CheckedFrom c, const Tensor& t, Device device) {
```
Should this belong here?
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorUtils.cpp

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorUtils.h

The file has `checkDeviceType` which only checks for device type but does not consider the index of the device.
However functions are marked with TORCH_API (not sure what this exactly implies) so not sure if we should add this :thinking:

kshitij12345 replied:
Reference to discuss: https://discuss.pytorch.org/t/what-is-torch-api/67398

mruberry replied:
Maybe. Let's think about improving our guidance for how ops should perform error checking separately, and then we can identify the best pattern. These device checks are typically simple to write, so I'm not sure a helper would add much.

kshitij12345 committed:
Merge branch 'master' into develop/numpy/add-take_along_dim

kshitij12345 committed:
Merge branch 'master' into develop/numpy/add-take_along_dim

kshitij12345 committed:
Merge branch 'master' into develop/numpy/add-take_along_dim

kshitij12345 committed:
remove trailing spaces

kshitij12345 committed:
skip test_out

kshitij12345 commented:
Gentle ping :)

kshitij12345 commented:
@mruberry Gentle Ping :)

mruberry commented on torch/_torch_docs.py:
```diff
@@ -7958,6 +7962,8 @@ def merge_dicts(*dicts):
 This is the second value returned by :meth:`torch.sort`.  See its documentation
 for the exact semantics of this method.
 
+See also :func:`torch.take_along_dim`.
```
Let's remove these "See also"s for now. I know I asked for a reference, but I think we should elaborate more and I don't want to block this PR on those changes.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8833,6 +8839,39 @@ def merge_dicts(*dicts):
+
+Selects values from :attr:`input` at the 1-dimensional indices from :attr:`indices` along the given :attr:`dim`.
+
+Functions returning index along dimension like torch.argmax and torch.argsort can be passed
```
"Functions that return indices along a dimension, like :func:\`torch.argmax\` and :func:\`torch.argsort\`, are designed to work with this function. See the examples below."

Make sure the function references work as expected.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8833,6 +8839,39 @@ def merge_dicts(*dicts):
+
+Example::
+
+    >>> t = torch.tensor([[10, 30, 20], [60, 40, 50]])
```
Nice examples.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -635,6 +635,35 @@ def sample_inputs_gather(op_info, device, dtype, requires_grad):
             )
 
+def sample_inputs_take_along_dim(op_info, device, dtype, requires_grad):
+    return (SampleInput((make_tensor((S, S), device, dtype,
```
These are going to need a tweak since SampleInput's "input" attribute must be a tensor now. So the pattern is:

```
SampleInput(make_tensor(...), args=(...))
```

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2653,6 +2682,15 @@ def _make_tensor_helper(shape, low=None, high=None):
+    OpInfo('take_along_dim',
+           dtypes=all_types_and_complex_and(torch.bool, torch.float16),
+           dtypesIfCUDA=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
+           test_inplace_grad=False,
```
supports_inplace_autograd=False

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2653,6 +2682,15 @@ def _make_tensor_helper(shape, low=None, high=None):
+           test_inplace_grad=False,
+           skips=(
+               # `take_along_dim` does not correctly warn when resizing out= inputs
+               SkipInfo('TestCommon', 'test_out'),
```
Seems like we should fix this?

mruberry reviewed (commented):
Thanks for the pings, @kshitij12345. Just a few clean-up / merge issues, although I am curious if we can just fix the out= behavior to get that test running. If we don't do it now we'll just have to do it later.

kshitij12345 committed:
Merge branch 'master' into develop/numpy/add-take_along_dim

kshitij12345 committed:
warning on resizing of output

kshitij12345 committed:
update docs

kshitij12345 committed:
update as per latest OpInfo structure

kshitij12345 commented:
Fixed the warning issue and removed the skip.

Verified that the links in document work fine.
Latest Doc Link: https://11739718-65600975-gh.circle-artifacts.com/0/docs/generated/torch.take_along_dim.html

Should be good to go.

Thanks!!

kshitij12345 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -808,7 +809,7 @@ Tensor index_fill(const Tensor & self, int64_t dim, const Tensor & index, const
 
 Tensor & gather_out_cpu_cuda(Tensor & result, const Tensor & self, int64_t dim, const Tensor & index, bool sparse_grad) {
-  result.resize_(index.sizes());
+  resize_output(result, index.sizes());
```
For reference -> fix for the `out=` test skip.

mruberry reviewed (approved):
Nice work, @kshitij12345! Great to have torch.take_along_dim() (finally!)!

mruberry commented:
What's interesting is that the ROCm failure appears to be related, but I'm not sure why it's complaining about the dispatch?

Would you take a look? If what's happening is a complete mystery we can call in one of our dispatch experts to help.

kshitij12345 commented:
I think it has to do with https://github.com/pytorch/pytorch/pull/54466.
(Wonder why only the ROCm build failed :thinking: )
Anyways will rebase on latest master and update the dispatch key. Mostly that should fix it.

kshitij12345 committed:
Merge branch 'master' into develop/numpy/add-take_along_dim

kshitij12345 committed:
update dispatch
